lib/charms/layer/kubernetes_common.py

```diff
@@ -733,14 +733,45 @@ def enable_ipv6_forwarding():
     check_call(["sysctl", "net.ipv6.conf.all.forwarding=1"])
 
 
-def get_bind_addrs(ipv4=True, ipv6=True):
-    """Get all global-scoped addresses that we might bind to."""
+def _as_interface(addr_str):
+    try:
+        return ipaddress.ip_interface(addr_str)
+    except ValueError:
+        return None
+
+
+def _get_bind_addrs_json():
+    try:
+        output = check_output(["ip", "-j", "-br", "addr", "show", "scope", "global"])
+    except CalledProcessError:
+        # stderr will have any details, and go to the log
+        hookenv.log("Unable to determine global addresses", hookenv.ERROR)
+        return []
+    addrs = json.loads(output.decode("utf8"))
+    for addr in addrs:
+        for ifc in addr["addr_info"]:
+            local, prefixlen = ifc.get("local"), ifc.get("prefixlen")
+            ifc["interface"] = _as_interface(f"{local}/{prefixlen}")
+    return addrs
+
+
+def _get_bind_addrs_table():
     try:
         output = check_output(["ip", "-br", "addr", "show", "scope", "global"])
     except CalledProcessError:
         # stderr will have any details, and go to the log
         hookenv.log("Unable to determine global addresses", hookenv.ERROR)
         return []
+    addrs = []
+    for line in output.decode("utf8").splitlines():
+        intf, state, *intf_addrs = line.split()
+        addr_info = [dict(interface=_as_interface(addr_str)) for addr_str in intf_addrs]
+        addrs.append(dict(ifname=intf, operstate=state, addr_info=addr_info))
+    return addrs
+
+
+def get_bind_addrs(ipv4=True, ipv6=True):
+    bind_addrs = _get_bind_addrs_json() or _get_bind_addrs_table()
```

Comment:
can we not rely solely on the json-based method for all cases?

Reply:
i believe that `series=bionic` doesn't support this `--json` output.. but i could be wrong

Reply:
OO... I stand corrected.  This does work on bionic.  It must be `xenial` where this doesn't work.  But we've recently dropped xenial soooo.... i could go for using the json version only

@Cynerva or @kwmonroe thoughts?

Reply:
this logic gets much easier when we don't have to worry about reading the table output from `ip addr show`

Reply:
Sounds good to me. :+1:

Reply:
ok, i'll rip out the parsing of the table, and only use the json formatted output